### PR TITLE
Sync should throw an error for any error in plugin discovery

### PR DESCRIPTION
**Pull request by chandrareddyp**
### What this PR does / why we need it

This pull request is to fix the UX message that appears when a plugin sync fails during `context creation`. Prior to this fix, any errors that occurred during plugin discovery were ignored, so the create context command would erroneously show that `All required plugins are already installed and up-to-date`, even though plugin discovery had actually failed. With this fix, the command will display the message `Unable to automatically sync the plugins from the target context. Please run the 'tanzu plugin sync' command to sync plugins manually`.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR
1. The E2E tests haven been updated
2. Tested locally
```
**Before this fix:**
❯ t  context create --name tmc12 --endpoint unstable22.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[!] unable to list plugin from discovery 'default-tmc12': API error, status code: 404
[i] All required plugins are already installed and up-to-date
❯ 
**After fix:**
❯ t  context create --name tmc24 --endpoint unstable22.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[!] unable to automatically sync the plugins from target context. Please run 'tanzu plugin sync' command to sync plugins manually, error: 'unable to list plugins from discovery source 'default-tmc24': API error, status code: 404'
❯ 
```

### Release note

```release-note

```

### CODE REVIEW

**Review by marckhouzam — COMMENTED** [review]
Thanks @chandrareddyp.  These "small" fixes are pretty complicated because they touch very sensitive code and multiple code paths, but I think you did a good job here.

Now that we report a proper error when the sync fails dugin context creation, should we also return an error code to show the command "failed"?
We would just need to return the error code here: https://github.com/vmware-tanzu/tanzu-cli/blob/1e5fa52eecf370c953955841a6b14a55bc9d3e8b/pkg/command/context.go#L145
For example:
```
$ tz  context create --name tmc12 --endpoint unstable22.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[x] unable to list plugin from discovery 'default-tmc12': API error, status code: 404
[!] unable to automatically sync the plugins from target context. Please run 'tanzu plugin sync' command to sync plugins manually
$ echo $?
0
```

**Comment by chandrareddyp** [comment]
> Now that we report a proper error when the sync fails dugin context creation, should we also return an error code to show the command "failed"?
> We would just need to return the error code here:

That's a good point. Yes, we can log and return the error in the syncContextPlugins() function. However, I don't think we need to process that error in the createCtx() or useCtx() methods, as their purpose is different. Just logging the error information is enough.

File: pkg/pluginmanager/manager.go [hunk 1]
```diff
@@ -87,24 +87,29 @@ func discoverPlugins(pd []configtypes.PluginDiscovery) ([]discovery.Discovered,
...
+			log.Errorf("unable to create discovery: %v", err.Error())
+			errorList = append(errorList, errors.Wrapf(err, "unable to create discovery"))
+			continue
 		}
 
 		plugins, err := discObject.List()
 		if err != nil {
-			log.Warningf("unable to list plugin from discovery '%v': %v", discObject.Name(), err.Error())
+			log.Errorf("unable to list plugin from discovery '%v': %v", discObject.Name(), err.Error())
+			errorList = append(errorList, errors.Wrapf(err, "unable to list plugin from discovery '%v'", discObject.Name()))
```
[the remark this answers is not in the document]
Comment (chandrareddyp) [reply pkg/pluginmanager/manager.go]: agree with you, better avoid logging any error/warning, let caller decide what they want to do with errors.

File: pkg/pluginmanager/manager.go [hunk 2]
```diff
@@ -87,24 +87,27 @@ func discoverPlugins(pd []configtypes.PluginDiscovery) ([]discovery.Discovered,
...
+// discoverSpecificPlugins returns all plugins that match the specified criteria from all PluginDiscovery sources,
+// along with an aggregated error (if any) that occurred while creating the plugin discovery source or fetching plugins.
 func discoverSpecificPlugins(pd []configtypes.PluginDiscovery, criteria *discovery.PluginDiscoveryCriteria) ([]discovery.Discovered, error) {
 	allPlugins := make([]discovery.Discovered, 0)
+	errorList := make([]error, 0)
 	for _, d := range pd {
 		discObject, err := discovery.CreateDiscoveryFromV1alpha1(d, criteria)
 		if err != nil {
-			return nil, errors.Wrapf(err, "unable to create discovery")
+			errorList = append(errorList, errors.Wrapf(err, "unable to create discovery source for '%v', error: '%v'", discObject.Name(), err.Error()))
```
Comment (marckhouzam) [inline pkg/pluginmanager/manager.go:98]: We are already wrapping the error through `errors.Wrapf` so we don't need to add it to the text of the message.
If you leave it like this it looks like:
```
tz context create --name tmc12 --endpoint unstable22.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[!] unable to automatically sync the plugins from target context. Please run 'tanzu plugin sync' command to sync plugins manually, error: 'unable to list plugins from discovery source 'default-tmc12', error: 'API error, status code: 404': API error, status code: 404'
```
Notice that `API error, status code: 404` is printed twice.
If you change it we get:
```
tz context create --name tmc13 --endpoint unstable22.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[!] unable to automatically sync the plugins from target context. Please run 'tanzu plugin sync' command to sync plugins manually, error: 'unable to list plugins from discovery source 'default-tmc13': API error, status code: 404'
```

Comment (chandrareddyp) [reply pkg/pluginmanager/manager.go]: that is good suggestion. here is error message after removing error info in `errors.Wrapf` :
```
❯ t  context create --name tmc24 --endpoint unstable22.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[!] unable to automatically sync the plugins from target context. Please run 'tanzu plugin sync' command to sync plugins manually, error: 'unable to list plugins from discovery source 'default-tmc24': API error, status code: 404'
```

File: test/e2e/framework/context_create_operations.go [hunk 3]
```diff
@@ -45,21 +45,22 @@ func (cc *contextCreateOps) CreateContextWithEndPoint(contextName, endpoint stri
...
 	log.Infof(ContextCreated, contextName)
-	return out.String(), err
+	return out.String(), stdErr.String(), err
 }
 
 func (cc *contextCreateOps) CreateContextWithKubeconfig(contextName, kubeconfigPath, kubeContext string, opts ...E2EOption) error {
 	createContextCmd := fmt.Sprintf(CreateContextWithKubeconfigFile, "%s", kubeconfigPath, kubeContext, contextName)
-	out, _, err := cc.cmdExe.TanzuCmdExec(createContextCmd, opts...)
+	out, stdErr, err := cc.cmdExe.TanzuCmdExec(createContextCmd, opts...)
+	fmt.Println(stdErr.String())
```
Comment (marckhouzam) [inline test/e2e/framework/context_create_operations.go:63]: Did you mean to leave this printout?

**Review by marckhouzam — APPROVED** [review]
LGTM
Thanks @chandrareddyp !

**Pull request merged**